src/core.rs

```diff
@@ -153,6 +155,73 @@ impl<'m, 'r> EncryptWithChosenRandomness<EncryptionKey, &'m RawPlaintext, &'r Pr
     }
 }
 
+impl<'m> Encrypt<DecryptionKey, &'m RawPlaintext, RawCiphertext> for Paillier {
+    fn encrypt(dk: &DecryptionKey, m: &'m RawPlaintext) -> RawCiphertext {
+        let (mp, mq) = crt_decompose(&m.0, &dk.pp, &dk.qq);
+        
+        let (cp, cq) = join(
+            || {
+                let rp = BigInt::sample_below(&dk.p);
+                let rnp = BigInt::modpow(&rp, &dk.n, &dk.pp);
+                let gmp = (1 + mp * &dk.n) % &dk.pp; // TODO[Morten] maybe there's more to get here
+                let cp = (gmp * rnp) % &dk.pp;
+                cp
+            },
+            || {
+                let rq = BigInt::sample_below(&dk.q);
+                let rnq = BigInt::modpow(&rq, &dk.n, &dk.qq);
+                let gmq = (1 + mq * &dk.n) % &dk.qq; // TODO[Morten] maybe there's more to get here
+                let cq = (gmq * rnq) % &dk.qq;
+                cq
+            }
+        );
+
+        // let rp = BigInt::sample_below(&dk.p);
+        // let rnp = BigInt::modpow(&rp, &dk.n, &dk.pp);
+        // let gmp = (1 + mp * &dk.n) % &dk.pp; // TODO[Morten] maybe there's more to get here
+        // let cp = (gmp * rnp) % &dk.pp;
+        
+        // let rq = BigInt::sample_below(&dk.q);
+        // let rnq = BigInt::modpow(&rq, &dk.n, &dk.qq);
+        // let gmq = (1 + mq * &dk.n) % &dk.qq; // TODO[Morten] maybe there's more to get here
+        // let cq = (gmq * rnq) % &dk.qq;
+
+        let c = crt_recombine(cp, cq, &dk.pp, &dk.qq, &dk.ppinv);
+        RawCiphertext(c)
+    }
+}
+
+impl<'m, 'r> EncryptWithChosenRandomness<DecryptionKey, &'m RawPlaintext, &'r Randomness, RawCiphertext> for Paillier {
+    fn encrypt_with_chosen_randomness(dk: &DecryptionKey, m: &'m RawPlaintext, r: &'r Randomness) -> RawCiphertext {
+        let (mp, mq) = crt_decompose(&m.0, &dk.pp, &dk.qq);
+        let (rp, rq) = crt_decompose(&r.0, &dk.pp, &dk.qq);
+        let (cp, cq) = join(
+            || {
+                let rnp = BigInt::modpow(&rp, &dk.n, &dk.pp);
+                let gmp = (1 + mp * &dk.n) % &dk.pp; // TODO[Morten] maybe there's more to get here
+                let cp = (gmp * rnp) % &dk.pp;
+                cp
+            },
+            || {
+                let rnq = BigInt::modpow(&rq, &dk.n, &dk.qq);
+                let gmq = (1 + mq * &dk.n) % &dk.qq; // TODO[Morten] maybe there's more to get here
+                let cq = (gmq * rnq) % &dk.qq;
+                cq
+            }
+        );
+        let c = crt_recombine(cp, cq, &dk.pp, &dk.qq, &dk.ppinv);
+        RawCiphertext(c)
+    }
+}
+
+impl<'m, 'r> EncryptWithChosenRandomness<DecryptionKey, &'m RawPlaintext, &'r PrecomputedRandomness, RawCiphertext> for Paillier {
+    fn encrypt_with_chosen_randomness(dk: &DecryptionKey, m: &'m RawPlaintext, rn: &'r PrecomputedRandomness) -> RawCiphertext {
+        let gm = (1 + &m.0 * &dk.n) % &dk.nn;
+        let c = (gm * &rn.0) % &dk.nn;
+        RawCiphertext(c)
```

Comment:
why not using crt here as well?

Reply:
I wasn't sure it was worth it since it's only two multiplications